File: solution/kotlin/src/test/kotlin/parse/dont/validate/NIRProperties.kt
package parse.dont.validate

import io.kotest.core.spec.style.StringSpec
import io.kotest.property.forAll
import parse.dont.validate.Generators.nirGenerator
import parse.dont.validate.NIR.Parser.parseNIR

class NIRProperties : StringSpec({
    "Nir round tripping" {
        forAll(nirGenerator) { validNIR ->
            parseNIR(validNIR.toString()).getOrNull() == validNIR
        }
    }
})

File: solution/scala/src/test/scala/parse/dont/validate/NIRProperties.scala
package parse.dont.validate

import org.scalacheck.Prop.forAll
import org.scalacheck.{Arbitrary, Gen, Properties}
import org.scalatest.EitherValues
import parse.dont.validate.Generators.validNIR
import parse.dont.validate.NIRDomain.*
import parse.dont.validate.NIRDomain.NIR.parseNIR

object NIRProperties extends Properties("NIR") with EitherValues {
  property("round tripping") = forAll(validNIR) { nir =>
    parseNIR(nir.toString).value == nir
  }
}
